src/macaron/repo_finder/provenance_finder.py

```diff
@@ -110,19 +117,89 @@ def find_npm_provenance(purl: PackageURL, npm_registry: NPMRegistry) -> InTotoPa
                 download_path = os.path.join(temp_dir, f"{artifact_id}.intoto.jsonl")
                 if not npm_provenance_asset.download(download_path):
                     logger.debug("Unable to find an npm provenance for %s@%s", artifact_id, version)
-                    return None
+                    return None, False
 
                 try:
                     # Load the provenance file.
                     provenance_payload = load_provenance_payload(download_path)
-                except LoadIntotoAttestationError as loadintotoerror:
-                    logger.error("Error while loading provenance %s", loadintotoerror)
-                    return None
 
-                return provenance_payload
+                    signed_download_path = f"{download_path}.signed"
+                    if not os.path.exists(signed_download_path):
+                        return provenance_payload, False
+
+                    # Compare to signed provenance.
```

Comment:
The provenances can be published in various places, e.g., `npm` registry, GitHub Attestations, GitHub releases, each requiring a different mechanism to verify, e.g., the comparison implemented in `_compare_unsigned_signed_provenance`, using SLSA verifier, etc. I'm not sure if the current implementation and design is flexible enough to allow extending for different verification methods. 

One design that I can think of is to separate discovery from verification. 

- The discovery function accepts a list of callables that discover a provenance, and it calls them one by one and stops as soon as it can discover a provenance. 
- The verification function accepts a provenance and a callable that implements the suitable verification. This function can call the verification method that is passed at the callsite. The verfication can also be turned off through the `.ini` configuration.

Using this approach, we can control at the callsite which discovery and verification functions are suitable for a specific `purl` type.

Please let me know what you think

Reply:
As per the updated PR description, finding and verifying accept callable functions to perform their respective tasks. A set of default functions are provided to retain the pre-existing functionality for npm and Gradle. The provenance verification can be disabled in the config.